restapi.go

```diff
@@ -41,42 +42,64 @@ var (
+// ID is the same as urlStr.
+func (s *Session) RequestWithContext(ctx context.Context, method, urlStr string, data interface{}) (response []byte, err error) {
+	return s.RequestWithContextBucketID(ctx, method, urlStr, data, strings.SplitN(urlStr, "?", 2)[0])
+}
+
+// RequestWithBucketID makes a (GET/POST/...) request to the Discord REST API
+// with JSON data.
 func (s *Session) RequestWithBucketID(method, urlStr string, data interface{}, bucketID string) (response []byte, err error) {
+	return s.RequestWithContextBucketID(context.TODO(), method, urlStr, data, bucketID)
+}
+
+// RequestWithContextBucketID makes a (GET/POST/...) request to the Discord
+// REST API with JSON data using the provided context.
+func (s *Session) RequestWithContextBucketID(ctx context.Context, method, urlStr string, data interface{}, bucketID string) (response []byte, err error) {
 	var body []byte
+
 	if data != nil {
 		body, err = json.Marshal(data)
 		if err != nil {
 			return
 		}
 	}
 
-	return s.request(method, urlStr, "application/json", body, bucketID, 0)
+	return s.request(ctx, method, urlStr, "application/json", body, bucketID, 0)
 }
 
-// request makes a (GET/POST/...) Requests to Discord REST API.
-// Sequence is the sequence number, if it fails with a 502 it will
-// retry with sequence+1 until it either succeeds or sequence >= session.MaxRestRetries
-func (s *Session) request(method, urlStr, contentType string, b []byte, bucketID string, sequence int) (response []byte, err error) {
+// request makes a (GET/POST/...) request to Discord REST API. Sequence is the
+// sequence number. If it fails with a 502 it will retry with sequence+1 until
+// it either succeeds or sequence >= session.MaxRestRetries.
+func (s *Session) request(ctx context.Context, method, urlStr, contentType string, b []byte, bucketID string, sequence int) (response []byte, err error) {
 	if bucketID == "" {
 		bucketID = strings.SplitN(urlStr, "?", 2)[0]
 	}
-	return s.RequestWithLockedBucket(method, urlStr, contentType, b, s.Ratelimiter.LockBucket(bucketID), sequence)
+
+	return s.RequestWithContextLockedBucket(ctx, method, urlStr, contentType, b, s.Ratelimiter.LockBucket(bucketID), sequence)
 }
 
-// RequestWithLockedBucket makes a request using a bucket that's already been locked
+// RequestWithLockedBucket makes a request using a bucket that's already been locked.
 func (s *Session) RequestWithLockedBucket(method, urlStr, contentType string, b []byte, bucket *Bucket, sequence int) (response []byte, err error) {
+	return s.RequestWithContextLockedBucket(context.TODO(), method, urlStr, contentType, b, bucket, sequence)
+}
+
+// RequestWithContextLockedBucket makes a request using a bucket that's already
+// been locked using the provided context.
+func (s *Session) RequestWithContextLockedBucket(ctx context.Context, method, urlStr, contentType string, b []byte, bucket *Bucket, sequence int) (response []byte, err error) {
```

Comment:
warning: cyclomatic complexity 23 of function (*Session).RequestWithContextLockedBucket() is high (> 15) (gocyclo)